Plugins/NativeEngine/Source/NativeEngine.cpp

```diff
@@ -1314,6 +1318,106 @@ namespace Babylon
+
+        // Create the output buffer if one wasn't passed in.
+        if (buffer.IsNull())
+        {
+            bufferOffset = 0;
+            bufferLength = targetTextureInfo.storageSize;
+            buffer = Napi::ArrayBuffer::New(env, bufferLength);
+        }
+
+        // Make sure the buffer is big enough to fit the output data.
+        if (buffer.ByteLength() < bufferOffset + bufferLength)
+        {
+            deferred.Reject(Napi::Error::New(env, "Provided buffer is too small.").Value());
+        }
+        else
+        {
+            bgfx::TextureHandle sourceTextureHandle{texture->Handle()};
+            auto tempTexture{false};
+
+            // If the image needs to be cropped (not starting at 0, or less than full width/height (accounting for requested mip level)),
+            // or if the texture was not created with the BGFX_TEXTURE_READ_BACK flag, then blit it to a temp texture.
+            if (x != 0 || y != 0 || width != (texture->Width() >> mipLevel) || height != (texture->Height() >> mipLevel) || (texture->Flags() & BGFX_TEXTURE_READ_BACK) == 0)
+            {
+                const bgfx::TextureHandle blitTextureHandle{bgfx::createTexture2D(width, height, /*hasMips*/ false, /*numLayers*/ 1, sourceTextureFormat, BGFX_TEXTURE_BLIT_DST | BGFX_TEXTURE_READ_BACK)};
+                bgfx::Encoder* encoder{GetUpdateToken().GetEncoder()};
+                encoder->blit(static_cast<uint16_t>(bgfx::getCaps()->limits.maxViews - 1), blitTextureHandle, /*dstMip*/ 0, /*dstX*/ 0, /*dstY*/ 0, /*dstZ*/ 0, sourceTextureHandle, mipLevel, x, y, /*srcZ*/ 0, width, height, /*depth*/ 0);
+
+                sourceTextureHandle = blitTextureHandle;
+                tempTexture = true;
+
+                // The requested mip level was blitted, so the source texture now has just one mip, so reset the mip level to 0.
+                mipLevel = 0;
+            }
+
+            // Allocate a buffer to store the source pixel data.
+            std::vector<uint8_t> textureBuffer(sourceTextureInfo.storageSize);
+
+            // Read the source texture.
+            m_graphicsContext.ReadTextureAsync(sourceTextureHandle, textureBuffer, mipLevel).then(arcana::inline_scheduler, *m_cancellationSource, [this, textureBuffer{std::move(textureBuffer)}, sourceTextureInfo, targetTextureInfo]() mutable {
+                // If the source texture format does not match the target texture format, convert it.
+                if (targetTextureInfo.format != sourceTextureInfo.format)
+                {
+                    std::vector<uint8_t> convertedTextureBuffer(targetTextureInfo.storageSize);
+                    bimg::imageConvert(&m_allocator, convertedTextureBuffer.data(), bimg::TextureFormat::Enum(targetTextureInfo.format), textureBuffer.data(), bimg::TextureFormat::Enum(sourceTextureInfo.format), sourceTextureInfo.width, sourceTextureInfo.height, /*depth*/ 1);
```

Comment:
Should probably check the result and throw exception if it failed.

Reply:
Updated. I'm throwing a regular C++ exception which I believe would be properly converted in the final continuation.